tests/common/devices.py

```diff
@@ -1301,6 +1301,122 @@ def execute (self, cmd) :
+                        raise ValueError("Trying to run module '{}' against asic_index '{}' on a single asic dut '{}'".format(self.attr, asic_index, self.sonichost.hostname))
+                return getattr(self.asics[asic_index], self.multi_asic_attr)(*module_args, **asic_complex_args)
+            elif type(asic_index) == str and asic_index.lower() == "all":
+                # All ASICs/namespace
+                if self.sonichost.facts['num_asic'] == 1:
+                    return [getattr(asic, self.multi_asic_attr)(*module_args, **asic_complex_args) for asic in self.asics]
+                return [getattr(asic, self.multi_asic_attr)(*module_args, **asic_complex_args) for asic in self.asics]
+            else:
+                raise ValueError("Argument 'asic_index' must be an int or string 'all'.")
+
+    def __getattr__(self, attr):
+        sonic_asic_attr = getattr(SonicAsic, attr, None)
+        if not attr.startswith("_") and sonic_asic_attr and callable(sonic_asic_attr):
+            self.multi_asic_attr = attr
+            return self._run_on_asics
+        else:
+            return getattr(self.sonichost, attr)  # For backward compatibility
+
+
+class DutHosts(object):
+    class _Nodes(list):
+
+        # Delegate the call to each of the nodes, return the results in a dict.
+        def _run_on_nodes(self, *module_args, **complex_args):
+            return {node.hostname: getattr(node, self.attr)(*module_args, **complex_args) for node in self}
+
+        # To support calling ansible modules on list of nodes.
+        def __getattr__(self, attr):
+            self.attr = attr
+            return self._run_on_nodes
+
+    def __init__(self, ansible_adhoc, tbinfo):
+        # TODO: Initialize the nodes in parallel using multi-threads?
+        self.nodes = self._Nodes([MultiAsicSonicHost(ansible_adhoc, hostname) for hostname in tbinfo["duts"]])
+        self.supervisor_nodes = self._Nodes([node for node in self.nodes if self._is_supervisor_node(node)])
+        self.frontend_nodes = self._Nodes([node for node in self.nodes if self._is_frontend_node(node)])
+
+
+    # To support operations like duthosts[0] and duthost['sonic1_hostname']
+    def __getitem__(self, index):
+        if type(index) == int:
+            return self.nodes[index]
+        elif type(index) == str:
+            for node in self.nodes:
+                if node.hostname == index:
+                    return node
+            raise KeyError("No node has hostname '{}'".format(index))
+        else:
+            raise IndexError("Bad index '{}'".format(index))
+
+    # To support iteration
+    def __iter__(self):
+        self._node_index = 0
+        return self.nodes.__iter__()
+
+    # To support iteration
+    def __next__(self):
+        if self._node_index < len(self.nodes):
+            node = self.nodes[self._node_index]
+            self._node_index += 1
+            return node
+        else:
+            raise StopIteration
```

Comment:
I realized that there is an easier and better way to make DutHosts iterable:

```python
    def __iter__(self):
        return iter(self.nodes)
```
Then there is no need to implement this `__next__` method.

BTW, it would be better to implement the `__len__` method as well:
```python
    def __len__(self):
        return len(self.nodes)
```

Reply:
Did the changes as mentioned above in the last commit. Also, added '__eq__' method to address one of the warnings above.

Reply:
Having trouble with __eq__ - got warnings - can you please help resolve.